File: backend_fastapi/app/routes/jobs.py
from fastapi import APIRouter, HTTPException, Depends, UploadFile, File, Form, Query, Body
from app.middleware.dependencies import get_current_user
from app.services.job_service import JobService
from typing import Optional, Any
import os, uuid, json
from datetime import datetime

router = APIRouter(prefix="/api/jobs", tags=["jobs"])
UPLOAD_DIR = "public/uploads/jobs"

@router.get("/employer/my-jobs")
async def get_my_jobs(current_user: dict = Depends(get_current_user)):
    try:
        if current_user["role"] not in ["employer", "admin"]:
            raise HTTPException(status_code=403, detail="Non autorisé")
        result = await JobService.get_employer_jobs(current_user["userId"])
        return result
    except Exception:
        raise HTTPException(status_code=500, detail="Erreur serveur")

@router.get("/employer/stats")
async def get_employer_stats(current_user: dict = Depends(get_current_user)):
    try:
        if current_user["role"] not in ["employer", "admin"]:
            raise HTTPException(status_code=403, detail="Non autorisé")
        stats = await JobService.get_employer_stats(current_user["userId"])
        return {"stats": stats}
    except Exception:
        raise HTTPException(status_code=500, detail="Erreur serveur")

@router.get("")
async def get_jobs(
    position: Optional[str] = Query(None),
    city: Optional[str] = Query(None),
    urgency: Optional[bool] = Query(None),
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
):
    try:
        result = await JobService.get_jobs(position, city, urgency, page, limit)
        return {
            "jobs": result["items"],
            "total": result["total"],
            "page": result["page"],
            "totalPages": result["pages"],
        }
    except Exception:
        raise HTTPException(status_code=500, detail="Erreur serveur")

@router.get("/{job_id}")
async def get_job(job_id: str):
    try:
        job = await JobService.get_job_by_id(job_id)
        return {"job": job}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception:
        raise HTTPException(status_code=500, detail="Erreur serveur")

@router.post("")
async def create_job(
    title: str = Form(...),
    cabinet: str = Form(...),
    description: str = Form(...),
    location: str = Form(...),
    address: Optional[str] = Form(None),
    contactEmail: Optional[str] = Form(None),
    salaryMin: Optional[str] = Form(None),
    salaryMax: Optional[str] = Form(None),
    contractType: str = Form(default="CDI"),
    urgency: Optional[str] = Form(default="false"),
    requirements: Optional[str] = Form(None),
    jobDescriptionFile: Optional[UploadFile] = File(None),
    current_user: dict = Depends(get_current_user),
):
    try:
        if current_user["role"] not in ["employer", "admin"]:
            raise HTTPException(status_code=403, detail="Non autorisé")

        file_name = None
        if jobDescriptionFile:
            os.makedirs(UPLOAD_DIR, exist_ok=True)
            ext = jobDescriptionFile.filename.split(".")[-1]
            file_name = f"job-{current_user['userId']}-{uuid.uuid4().hex}.{ext}"
            content = await jobDescriptionFile.read()
            with open(os.path.join(UPLOAD_DIR, file_name), "wb") as f:
                f.write(content)

        req_list = []
        if requirements:
            try:
                req_list = json.loads(requirements)
            except Exception:
                req_list = [r.strip() for r in requirements.split(",") if r.strip()]

        job_data = {
            "title": title,
            "cabinet": cabinet,
            "description": description,
            "requirements": req_list,
            "location": location,
            "address": address,
            "contactEmail": contactEmail,
            "salary": {
                "min": int(salaryMin) if salaryMin else 0,
                "max": int(salaryMax) if salaryMax else 0,
            },
            "contractType": contractType,
            "urgency": urgency in ("true", "1", "True"),
        }

        job = await JobService.create_job(current_user["userId"], job_data, file_name)
        return {"message": "Offre soumise — en attente de validation (24-48h)", "job": job}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail="Erreur serveur")

@router.put("/{job_id}")
async def update_job(
    job_id: str,
    request_body: dict = Body(...),
    current_user: dict = Depends(get_current_user),
):
    try:
        job = await JobService.update_job(job_id, request_body, current_user["userId"], current_user["role"])
        return {"message": "Offre mise à jour", "job": job}
    except ValueError as e:
        code = 403 if "authorized" in str(e).lower() else 404
        raise HTTPException(status_code=code, detail=str(e))
    except Exception:
        raise HTTPException(status_code=500, detail="Erreur serveur")

@router.put("/{job_id}/pipeline")
async def update_pipeline_stage(
    job_id: str,
    body: dict = Body(...),
    current_user: dict = Depends(get_current_user),
):
    try:
        stage = body.get("stage") or body.get("pipelineStage")
        if not stage:
            raise HTTPException(status_code=400, detail="stage requis")
        job = await JobService.update_pipeline_stage(job_id, stage, current_user["userId"], current_user["role"])
        return {"message": "Stage mis à jour", "job": job}
    except ValueError as e:
        code = 403 if "authorized" in str(e).lower() else 400
        raise HTTPException(status_code=code, detail=str(e))
    except HTTPException:
        raise
    except Exception:
        raise HTTPException(status_code=500, detail="Erreur serveur")

@router.delete("/{job_id}")
async def delete_job(job_id: str, current_user: dict = Depends(get_current_user)):
    try:
        result = await JobService.delete_job(job_id, current_user["userId"], current_user["role"])
        return result
    except ValueError as e:
        code = 403 if "authorized" in str(e).lower() else 404
        raise HTTPException(status_code=code, detail=str(e))
    except Exception:
        raise HTTPException(status_code=500, detail="Erreur serveur")

File: backend/controls/jobs.js
const Job = require("../model/job");
const Application = require("../model/application");
const AuditLog = require("../model/auditLog");
const sgMail = require("@sendgrid/mail");
require("dotenv").config();

sgMail.setApiKey(process.env.SENDGRID_API_KEY);

// ─── Helper: Audit Log ────────────────────────────────────────────────────────

const logAction = async (action, req, targetType, targetId, targetLabel, metadata = {}) => {
  try {
    await AuditLog.create({
      action,
      actor: req.userId,
      actorEmail: req.userEmail,
      actorRole: req.userRole,
      targetType,
      targetId,
      targetLabel,
      metadata,
      ipAddress: req.ip,
    });
  } catch (err) {
    console.error("AuditLog error:", err.message);
  }
};

// ─── GET /api/jobs — Offres validées avec filtres ─────────────────────────────

exports.getJobs = async (req, res) => {
  try {
    const { position, city, urgency, page = 1, limit = 20 } = req.query;

    const filter = { isValidated: true, status: "open" };
    if (position) filter.title = { $regex: position, $options: "i" };
    if (city) filter.location = { $regex: city, $options: "i" };
    if (urgency === "true") filter.urgency = true;

    const skip = (parseInt(page) - 1) * parseInt(limit);
    const [jobs, total] = await Promise.all([
      Job.find(filter)
        .populate("employer", "name company")
        .sort({ urgency: -1, createdAt: -1 })
        .skip(skip)
        .limit(parseInt(limit)),
      Job.countDocuments(filter),
    ]);

    res.json({ jobs, total, page: parseInt(page), totalPages: Math.ceil(total / parseInt(limit)) });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Erreur serveur" });
  }
};

// ─── GET /api/jobs/my-jobs — Offres de l'employeur connecté ──────────────────

exports.getMyJobs = async (req, res) => {
  try {
    const jobs = await Job.find({ employer: req.userId })
      .sort({ createdAt: -1 });
    res.json({ jobs });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Erreur serveur" });
  }
};

// ─── GET /api/jobs/:id ────────────────────────────────────────────────────────

exports.getJob = async (req, res) => {
  try {
    const job = await Job.findById(req.params.id).populate("employer", "name company email");
    if (!job) return res.status(404).json({ message: "Offre introuvable" });
    res.json({ job });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Erreur serveur" });
  }
};

// ─── POST /api/jobs — Créer une offre (employeur) ────────────────────────────

exports.createJob = async (req, res) => {
  try {
    const {
      title, cabinet, description, requirements,
      location, address, contactEmail,
      salaryMin, salaryMax, contractType, urgency,
    } = req.body;

    const jobData = {
      title,
      cabinet,
      description,
      requirements: requirements ? JSON.parse(requirements) : [],
      location,
      address,
      contactEmail,
      salary: {
        min: salaryMin ? Number(salaryMin) : undefined,
        max: salaryMax ? Number(salaryMax) : undefined,
      },
      contractType: contractType || "CDI",
      urgency: urgency === "true" || urgency === true,
      employer: req.userId,
      pipelineStage: "validating",
      isValidated: false,
    };

    // Fichier fiche de poste
    if (req.file) {
      jobData.jobDescriptionFile = req.file.filename;
    }

    const job = new Job(jobData);
    await job.save();

    await logAction("JOB_CREATED", req, "Job", job._id, `${title} - ${cabinet}`, { location, urgency });

    res.status(201).json({ message: "Offre soumise — en attente de validation (24-48h)", job });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Erreur serveur" });
  }
};

// ─── PUT /api/jobs/:id — Modifier une offre ───────────────────────────────────

exports.updateJob = async (req, res) => {
  try {
    const job = await Job.findById(req.params.id);
    if (!job) return res.status(404).json({ message: "Offre introuvable" });

    if (job.employer.toString() !== req.userId && req.userRole !== "admin") {
      return res.status(403).json({ message: "Non autorisé" });
    }

    // Champs non modifiables par l'employeur
    const forbidden = ["employer", "isValidated", "validatedAt", "validatedBy"];
    const updates = { ...req.body };
    if (req.userRole !== "admin") forbidden.forEach((f) => delete updates[f]);

    Object.assign(job, updates);
    await job.save();

    await logAction("JOB_UPDATED", req, "Job", job._id, job.title, { updates: Object.keys(updates) });

    res.json({ message: "Offre mise à jour", job });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Erreur serveur" });
  }
};

// ─── DELETE /api/jobs/:id ─────────────────────────────────────────────────────

exports.deleteJob = async (req, res) => {
  try {
    const job = await Job.findById(req.params.id);
    if (!job) return res.status(404).json({ message: "Offre introuvable" });

    if (job.employer.toString() !== req.userId && req.userRole !== "admin") {
      return res.status(403).json({ message: "Non autorisé" });
    }

    await Job.findByIdAndDelete(req.params.id);
    await logAction("JOB_DELETED", req, "Job", job._id, job.title);

    res.json({ message: "Offre supprimée" });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Erreur serveur" });
  }
};

// ─── PUT /api/jobs/pipeline/:id — Mettre à jour le stage pipeline ─────────────

exports.updatePipelineStage = async (req, res) => {
  try {
    const { stage } = req.body;
    const validStages = ["validating", "active", "sourcing", "interview", "closed"];
    if (!validStages.includes(stage)) {
      return res.status(400).json({ message: "Stage invalide" });
    }

    const job = await Job.findById(req.params.id);
    if (!job) return res.status(404).json({ message: "Offre introuvable" });

    if (job.employer.toString() !== req.userId && req.userRole !== "admin") {
      return res.status(403).json({ message: "Non autorisé" });
    }

    const oldStage = job.pipelineStage;
    job.pipelineStage = stage;
    if (stage === "closed") job.status = "closed";
    await job.save();

    await logAction("JOB_STAGE_CHANGED", req, "Job", job._id, job.title, { from: oldStage, to: stage });

    res.json({ message: "Stage mis à jour", job });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Erreur serveur" });
  }
};

// ─── GET /api/jobs/employer/stats ─────────────────────────────────────────────

exports.getEmployerStats = async (req, res) => {
  try {
    const jobs = await Job.find({ employer: req.userId }).select("_id");
    const jobIds = jobs.map((j) => j._id);

    const [activeOffers, totalApplications, pendingReview, interviews] = await Promise.all([
      Job.countDocuments({ employer: req.userId, status: "open", isValidated: true }),
      Application.countDocuments({ job: { $in: jobIds } }),
      Application.countDocuments({ job: { $in: jobIds }, employerStage: "validating" }),
      Application.countDocuments({ job: { $in: jobIds }, employerStage: "interview" }),
    ]);

    res.json({
      stats: {
        activeOffers,
        totalApplications,
        pendingReview,
        interviews,
      },
    });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Erreur serveur" });
  }
};
